### PR TITLE
Allow matching with prerelease versions

### LINKED ISSUE #504: Snaps fail to load when the only version has a prerelease tag
When the Snap has only one version, and it has a pre-release tag, such as `1.0.0-beta.1`, Flask fails to load the Snap with an errror:  
`Error when adding snap. Error: Failed to find a matching version in npm metadata for package "snap-sem-ver-error" and requested semver range "*"`

The loading works correctly when the version is specified directly
```typescript
    async function connect () {
      await ethereum.request({
        method: 'wallet_enable',
        params: [{
          wallet_snap: { [snapId]: {
            version: '1.0.0-beta.1'
          } },
        }]
      })
    }
```

Reproduction: https://github.com/piotr-roslaniec/snap-sem-ver-error

@piotr-roslaniec

### PR DESCRIPTION
**Pull request by FrederikBolding**
Fixes #504 

Allows version ranges to include prereleases, moves more semver handling to tested utility functions, use prerelease versions if no other matching versions are found.

### CODE REVIEW

File: packages/controllers/src/snaps/utils.test.ts [hunk 1]
```diff
@@ -122,3 +126,92 @@ describe('getSnapPrefix', () => {
...
+    expect(gtVersion('2.0.0', '1.0.0')).toBe(true);
+    expect(gtVersion('1.0.0', '1.2.3')).toBe(false);
+    expect(gtVersion('1.0.0', '2.0.0')).toBe(false);
+  });
+
+  it('supports pre-release versions', () => {
+    expect(gtVersion('1.0.0-beta.2', '1.0.0-beta.1')).toBe(true);
+    expect(gtVersion('1.0.0-beta.2', '1.2.3')).toBe(false);
+    expect(gtVersion('1.0.0', '1.0.0-beta.2')).toBe(true);
+    expect(gtVersion('1.2.3-beta.1', '1.0.0')).toBe(true);
```
Comment (ritave) [inline packages/controllers/src/snaps/utils.test.ts:166]: what happens in `gtVersion('1.2.3-beta.1', '1.2.3-alpha.2')`?

File: packages/controllers/src/snaps/utils.test.ts [hunk 2]
```diff
@@ -122,3 +126,92 @@ describe('getSnapPrefix', () => {
...
+  });
+
+  it('supports pre-release versions', () => {
+    expect(getTargetVersion(['1.0.0-beta.1'], '*')).toBe('1.0.0-beta.1');
+    expect(getTargetVersion(['1.0.0-beta.1', '1.0.0-beta.2'], '*')).toBe(
+      '1.0.0-beta.2',
+    );
+
+    expect(
+      getTargetVersion(['1.0.0-beta.1', '1.0.0-beta.2'], '^1.0.0-beta.1'),
```
Comment (ritave) [inline packages/controllers/src/snaps/utils.test.ts:182]: what happens in `getTargetVersion(['1.0.0-alpha.2', '1.0.0-beta.1'], '*');`?

File: packages/controllers/src/snaps/utils.test.ts [hunk 3]
```diff
@@ -122,3 +126,92 @@ describe('getSnapPrefix', () => {
...
+
+  it('doesnt return pre-release versions by default', () => {
+    expect(getTargetVersion(['1.0.0-beta.1', '1.0.0', '1.2.3'], '*')).toBe(
+      '1.2.3',
+    );
+
+    expect(getTargetVersion(['1.0.0-beta.1', '1.0.0', '1.2.3'], '^1.0.0')).toBe(
+      '1.2.3',
+    );
+    expect(getTargetVersion(['1.0.0-beta.1', '1.0.0'], '*')).toBe('1.0.0');
```
Comment (ritave) [inline packages/controllers/src/snaps/utils.test.ts:194]: what happens in `getTargetVersion(['0.9.0', '1.0.0-alpha.0'], '*')`?

**Review by ritave — CHANGES_REQUESTED** [review]

File: packages/controllers/src/snaps/utils.test.ts (hunk 1 above)
Comment (FrederikBolding) [reply packages/controllers/src/snaps/utils.test.ts:166]: `true`

File: packages/controllers/src/snaps/utils.test.ts (hunk 2 above)
Comment (FrederikBolding) [reply packages/controllers/src/snaps/utils.test.ts:182]: `1.0.0-beta.1`

File: packages/controllers/src/snaps/utils.test.ts (hunk 3 above)
Comment (FrederikBolding) [reply packages/controllers/src/snaps/utils.test.ts:194]: `0.9.0`

File: packages/controllers/src/snaps/utils.test.ts [hunk 4]
```diff
@@ -122,3 +126,92 @@ describe('getSnapPrefix', () => {
...
+    expect(satifiesVersionRange('1.2.3', '1.0.0')).toBe(false);
+  });
+
+  it('supports non-exact version ranges', () => {
+    expect(satifiesVersionRange('1.2.3', '^1.0.0')).toBe(true);
+    expect(satifiesVersionRange('2.0.0', '^1.0.0')).toBe(false);
+  });
+
+  it('prereleases can satisfy version range', () => {
+    expect(satifiesVersionRange('1.0.0-beta.1', '*')).toBe(true);
```
[the remark this answers is not in the document]
Comment (FrederikBolding) [reply packages/controllers/src/snaps/utils.test.ts:215]: `false`

**Review by ritave — APPROVED** [review]

**Pull request merged**